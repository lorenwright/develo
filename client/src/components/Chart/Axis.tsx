import {useMemo} from "react";
import * as React from "react";
import {ScaleLinear} from "d3";
import {Settings} from "../../types/ChartTypes";

type Props = {
  settings: Settings,
  xScale: ScaleLinear<any, any>,
  yScale: ScaleLinear<any, any>,
  chartWidth: number
}

const Axis: React.FC<Props> = ({settings, xScale, yScale, chartWidth}) => {
  /**
   * Get the tick marks for x-axis
   * @returns The ticks for x-axis
   */
  const xTicks = useMemo(() => {
    return xScale.ticks()
      .map(value => ({
        value,
        xOffset: xScale(value)
      }))
  }, [settings, chartWidth])

  /**
   * Get the tick marks for y-axis
   * @returns The ticks for y-axis
   */
  const yTicks = useMemo(() => {
    return yScale.ticks()
      .map(value => ({
        value,
        yOffset: yScale(value)
      }))
  }, [settings, chartWidth])
  return (
    <>
      <g>
        <path
          d={`M ${settings.marginLeft} ${settings.height - settings.marginBottom} H ${chartWidth - settings.marginLeft}`}
          stroke="#fff"
        />

        {xTicks.map(({ value, xOffset }) => (
          <g
            key={value}
            transform={`translate(${xOffset}, ${settings.height - settings.marginBottom})`}
          >
            <line
              y2="6"
              stroke="#fff"
            />
            <text
              key={value}
              style={{
                fontSize: "10px",
                textAnchor: "middle",
                transform: `translateY(20px)`,
                fill: '#fff'
              }}
            >
              { value }
            </text>
          </g>
        ))}
      </g>

      <g>
        <path
          d={`M ${settings.marginLeft} ${settings.marginTop} V ${settings.height - settings.marginBottom}`}
          stroke="#fff"
        />

        {yTicks.map(({ value, yOffset }) => (
          <g
            key={value}
            transform={`translate(${settings.marginLeft - 6}, ${yOffset})`}
          >
            <line
              x2="6"
              stroke="#fff"
            />
            <text
              key={value}
              style={{
                fontSize: "10px",
                textAnchor: "end",
                transform: `translate(-3px, 3px)`,
                fill: '#fff'
              }}
            >
              { value }
            </text>
          </g>
        ))}
      </g>
    </>
  )
}

export default Axis