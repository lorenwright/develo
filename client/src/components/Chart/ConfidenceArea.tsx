import * as React from "react";
import {useRef, useEffect} from "react";
import * as d3 from "d3";
import {ScaleLinear} from "d3";
import {Settings} from "../../types/ChartTypes";

type Props = {
  settings: Settings,
  rangeData: object[],
  xScale: ScaleLinear<any, any>,
  yScale: ScaleLinear<any, any>,
  chartWidth: number
}

const ConfidenceArea: React.FC<Props> = ({settings, rangeData, xScale, yScale, chartWidth}) => {

  const ref = useRef<SVGPathElement>(null)

  useEffect(() => {
    // The confidence area, determined by y0 (bottom y value) and y1 (top y value) for x
    const area = d3.area()
      .x((d:any) => xScale(d[settings.xAttribute]))
      .y0((d:any) => yScale(d[settings.yBottomAttribute]))
      .y1((d:any) => yScale(d[settings.yTopAttribute]))
      .curve(d3.curveLinear)

    // Add the confidence area to chart
    d3.select(ref.current)
      .datum(rangeData)
      .attr("fill", settings.confidenceAreaFill)
      .attr("stroke", settings.confidenceAreaStroke)
      .attr('d',
        // @ts-ignore
        area
      )
  }, [rangeData, chartWidth])

  return (
    <path ref={ref} />
  )
}

export default ConfidenceArea;