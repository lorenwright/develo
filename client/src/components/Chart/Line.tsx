import * as React from "react";
import {useRef, useEffect, useState} from "react";
import * as d3 from "d3";
import {ScaleLinear} from "d3";
import {Settings} from "../../types/ChartTypes";

type Props = {
  settings: Settings,
  data: object[],
  xScale: ScaleLinear<any, any>,
  yScale: ScaleLinear<any, any>,
  yAttribute: string,
  animate: boolean,
  chartWidth: number,
  showLabel: boolean
}

const Line: React.FC<Props> = ({settings, data, xScale, yScale, yAttribute, animate = false, chartWidth, showLabel}) => {
  const [labelDims, setLabelDims] = useState([])
  const path = useRef<SVGPathElement>(null)

  useEffect(() => {
    // The line we want to draw
    const line = d3.line()
      .x((d:any) => xScale(d[settings.xAttribute]))
      .y((d:any) => yScale(d[yAttribute]))
      .curve(d3.curveLinear)

    // Add the line to the chart
    d3.select(path.current)
      .datum(data)
      .attr('stroke', settings.lineStroke)
      .attr('stroke-width', settings.lineStrokeWidth)
      .attr('fill', 'none')
      .attr('d',
        // @ts-ignore
        line
      )

    // If showLabel is set, then we need to set the dimensions of where the label will be displayed
    if (data && data[data.length - 1] && showLabel) {
      setLabelDims([
        // @ts-ignore
        xScale(data[data.length - 1][settings.xAttribute] - 10), yScale(data[data.length - 1][yAttribute])
      ])
    }
  }, [data, chartWidth, settings, xScale, yScale, showLabel, yAttribute])

  return (
    <g>
      <path ref={path} className={`${animate ? 'draw-line' : ''}`} />
      {showLabel && labelDims.length > 0 && (
        <text transform={`translate(${labelDims[0]}, ${labelDims[1]})`} fill={settings.lineStroke} fontSize={10}>{yAttribute}</text>
      )}
    </g>
  )
}

export default Line;