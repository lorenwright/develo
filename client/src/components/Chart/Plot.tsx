import * as React from "react";
import {useRef} from "react";
import {ScaleLinear} from "d3";
import {Settings} from "../../types/ChartTypes";

type Props = {
  settings: Settings,
  data: object[],
  xScale: ScaleLinear<any, any>,
  yScale: ScaleLinear<any, any>,
  yAttribute: string,
  chartWidth: number
}

const Plot: React.FC<Props> = ({settings, data, xScale, yScale, yAttribute, chartWidth}) => {
  const ref = useRef<SVGGElement>(null)

  /**
   * The points that we should be rendering based off of x & y
   * @type {array} [x, y]
   */
  const points = data.map((d:any) => {
    return [d[settings.xAttribute], d[yAttribute]]
  })

  return (
    <g ref={ref}>
      {points.map((p:number[], idx) => (
        <circle key={idx} className="plot" cx={xScale(p[0])} cy={yScale(p[1])} r={6} fill={settings.lineStroke} stroke="#fff" />
      ))}
    </g>
  )
}

export default Plot;