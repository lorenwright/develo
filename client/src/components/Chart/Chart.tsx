import Axis from "./Axis";
import * as React from "react";
import ConfidenceArea from "./ConfidenceArea";
import * as d3 from "d3";
import {useEffect, useMemo, useRef, useState} from "react";
import {Settings} from "../../types/ChartTypes";
import Line from "./Line";
import Plot from "./Plot";

type Props = {
  settings: Settings,
  rangeData: object[],
  patientData: object[],
  label: string
}
const Chart: React.FC<Props> = ({settings, rangeData, patientData, label}) => {

  const [chartWidth, setChartWidth] = useState(0)
  const ref = useRef<HTMLDivElement>(null)

  useEffect(() => {
    getChartWidth()
    window.addEventListener('resize', getChartWidth)
  }, [])

  /**
   * Get the width of the chart based off of parent container
   */
  const getChartWidth = () => {
    // @ts-ignore
    const newWidth = ref.current.clientWidth - 16 // 16 is for padding
    setChartWidth(newWidth);
  }

  /**
   * Set the xScale for d3 chart
   * @returns The xScale
   */
  const xScale = useMemo(() => {
    return d3.scaleLinear()
      .domain([settings.xMin, settings.xMax])
      .range([settings.marginLeft, chartWidth - settings.marginLeft])
  }, [settings, chartWidth])

  /**
   * Set the yScale for d3 chart
   * @returns The yScale
   */
  const yScale = useMemo(() => {
    return d3.scaleLinear()
      .domain([settings.yMax, settings.yMin])
      .range([settings.marginTop, settings.height - settings.marginBottom])
  }, [settings, chartWidth])

  return (
    <div className="bg-slate-700/25 rounded-xl p-4" ref={ref}>
      <h2 className="text-2xl text-sky-500 font-bold text-center">{label}</h2>
      <div className="flex items-center">
        <div className="w-3">
          <h4 className="text-xs text-sky-500 text-center -rotate-90">Measurement</h4>
        </div>
        <div>
          <svg className="" width={chartWidth} height={settings.height}>
            <Axis settings={settings} xScale={xScale} yScale={yScale} chartWidth={chartWidth} />
            <ConfidenceArea settings={settings} rangeData={rangeData} xScale={xScale} yScale={yScale} chartWidth={chartWidth} />
            {settings.percentileLines.map((line, idx) => (
                <Line key={idx} settings={settings} data={rangeData} xScale={xScale} yScale={yScale} yAttribute={line} animate={false} chartWidth={chartWidth} showLabel />
            ))}
            <Line settings={{...settings, lineStroke: 'black', lineStrokeWidth: '2'}} data={patientData} xScale={xScale} yScale={yScale} yAttribute="value" animate chartWidth={chartWidth} showLabel={false} />
            <Plot settings={{...settings}} data={patientData} xScale={xScale} yScale={yScale} yAttribute="value" chartWidth={chartWidth} />
          </svg>
          <h4 className="text-xs text-sky-500 text-center mt-1">Age (in months)</h4>
        </div>
      </div>
    </div>
  )
}

export default Chart;