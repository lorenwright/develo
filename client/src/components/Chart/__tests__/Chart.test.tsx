import * as renderer from "react-test-renderer";
import Chart from '../Chart';
import '@testing-library/jest-dom';
import React from "react";

describe('Chart', () => {
  it('renders correctly', () => {

    const patient = {
      sex: 1,
      weight: [
        {
          Agemos: 33,
          value: 12
        },
        {
          Agemos: 50,
          value: 20
        },
        {
          Agemos: 70,
          value: 25
        },
        {
          Agemos: 95,
          value: 29
        },
        {
          Agemos: 120,
          value: 47
        },
        {
          Agemos: 170,
          value: 70
        },
        {
          Agemos: 200,
          value: 78
        },
        {
          Agemos: 230,
          value: 90
        }
      ]
    }

    const defaultSettings = {
      height: 700,
      marginLeft: 30,
      marginRight: 0,
      marginTop: 5,
      marginBottom: 20,
      yMax: 100,
      yMin: 0,
      xMin: 0,
      xMax: 100,
      confidenceAreaFill: '#bae6fd',
      confidenceAreaStroke: 'none',
      lineStroke: '#38bdf8',
      lineStrokeWidth: '2',
      yTopAttribute: 'P97',
      yBottomAttribute: 'P3',
      xAttribute: 'Agemos',
      percentileLines: ['P3', 'P25', 'P50', 'P75', 'P97']
    }

    const range = [
      {P3: 0, P25: 1, P50: 2, P75: 3, P97: 4},
      {P3: 3, P25: 25, P50: 50, P75: 75, P97: 97},
    ]

    const component = renderer.create(
        <Chart settings={defaultSettings} rangeData={range} patientData={patient.weight} label="Weight for Age" />
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  })
})