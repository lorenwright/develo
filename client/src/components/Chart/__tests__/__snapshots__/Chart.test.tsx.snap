// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chart renders correctly 1`] = `
<div
  className="bg-slate-700/25 rounded-xl p-4"
>
  <h2
    className="text-2xl text-sky-500 font-bold text-center"
  >
    Weight for Age
  </h2>
  <div
    className="flex items-center"
  >
    <div
      className="w-3"
    >
      <h4
        className="text-xs text-sky-500 text-center -rotate-90"
      >
        Measurement
      </h4>
    </div>
    <div>
      <svg
        className=""
        height={700}
        width={0}
      >
        <g>
          <path
            d="M 30 680 H -30"
            stroke="#fff"
          />
          <g
            transform="translate(30, 680)"
          >
            <line
              stroke="#fff"
              y2="6"
            />
            <text
              style={
                {
                  "fill": "#fff",
                  "fontSize": "10px",
                  "textAnchor": "middle",
                  "transform": "translateY(20px)",
                }
              }
            >
              0
            </text>
          </g>
          <g
            transform="translate(24, 680)"
          >
            <line
              stroke="#fff"
              y2="6"
            />
            <text
              style={
                {
                  "fill": "#fff",
                  "fontSize": "10px",
                  "textAnchor": "middle",
                  "transform": "translateY(20px)",
                }
              }
            >
              10
            </text>
          </g>
          <g
            transform="translate(18, 680)"
          >
            <line
              stroke="#fff"
              y2="6"
            />
            <text
              style={
                {
                  "fill": "#fff",
                  "fontSize": "10px",
                  "textAnchor": "middle",
                  "transform": "translateY(20px)",
                }
              }
            >
              20
            </text>
          </g>
          <g
            transform="translate(12, 680)"
          >
            <line
              stroke="#fff"
              y2="6"
            />
            <text
              style={
                {
                  "fill": "#fff",
                  "fontSize": "10px",
                  "textAnchor": "middle",
                  "transform": "translateY(20px)",
                }
              }
            >
              30
            </text>
          </g>
          <g
            transform="translate(6, 680)"
          >
            <line
              stroke="#fff"
              y2="6"
            />
            <text
              style={
                {
                  "fill": "#fff",
                  "fontSize": "10px",
                  "textAnchor": "middle",
                  "transform": "translateY(20px)",
                }
              }
            >
              40
            </text>
          </g>
          <g
            transform="translate(0, 680)"
          >
            <line
              stroke="#fff"
              y2="6"
            />
            <text
              style={
                {
                  "fill": "#fff",
                  "fontSize": "10px",
                  "textAnchor": "middle",
                  "transform": "translateY(20px)",
                }
              }
            >
              50
            </text>
          </g>
          <g
            transform="translate(-6, 680)"
          >
            <line
              stroke="#fff"
              y2="6"
            />
            <text
              style={
                {
                  "fill": "#fff",
                  "fontSize": "10px",
                  "textAnchor": "middle",
                  "transform": "translateY(20px)",
                }
              }
            >
              60
            </text>
          </g>
          <g
            transform="translate(-11.999999999999998, 680)"
          >
            <line
              stroke="#fff"
              y2="6"
            />
            <text
              style={
                {
                  "fill": "#fff",
                  "fontSize": "10px",
                  "textAnchor": "middle",
                  "transform": "translateY(20px)",
                }
              }
            >
              70
            </text>
          </g>
          <g
            transform="translate(-18, 680)"
          >
            <line
              stroke="#fff"
              y2="6"
            />
            <text
              style={
                {
                  "fill": "#fff",
                  "fontSize": "10px",
                  "textAnchor": "middle",
                  "transform": "translateY(20px)",
                }
              }
            >
              80
            </text>
          </g>
          <g
            transform="translate(-24, 680)"
          >
            <line
              stroke="#fff"
              y2="6"
            />
            <text
              style={
                {
                  "fill": "#fff",
                  "fontSize": "10px",
                  "textAnchor": "middle",
                  "transform": "translateY(20px)",
                }
              }
            >
              90
            </text>
          </g>
          <g
            transform="translate(-30, 680)"
          >
            <line
              stroke="#fff"
              y2="6"
            />
            <text
              style={
                {
                  "fill": "#fff",
                  "fontSize": "10px",
                  "textAnchor": "middle",
                  "transform": "translateY(20px)",
                }
              }
            >
              100
            </text>
          </g>
        </g>
        <g>
          <path
            d="M 30 5 V 680"
            stroke="#fff"
          />
          <g
            transform="translate(24, 5)"
          >
            <line
              stroke="#fff"
              x2="6"
            />
            <text
              style={
                {
                  "fill": "#fff",
                  "fontSize": "10px",
                  "textAnchor": "end",
                  "transform": "translate(-3px, 3px)",
                }
              }
            >
              100
            </text>
          </g>
          <g
            transform="translate(24, 72.49999999999999)"
          >
            <line
              stroke="#fff"
              x2="6"
            />
            <text
              style={
                {
                  "fill": "#fff",
                  "fontSize": "10px",
                  "textAnchor": "end",
                  "transform": "translate(-3px, 3px)",
                }
              }
            >
              90
            </text>
          </g>
          <g
            transform="translate(24, 139.99999999999997)"
          >
            <line
              stroke="#fff"
              x2="6"
            />
            <text
              style={
                {
                  "fill": "#fff",
                  "fontSize": "10px",
                  "textAnchor": "end",
                  "transform": "translate(-3px, 3px)",
                }
              }
            >
              80
            </text>
          </g>
          <g
            transform="translate(24, 207.50000000000003)"
          >
            <line
              stroke="#fff"
              x2="6"
            />
            <text
              style={
                {
                  "fill": "#fff",
                  "fontSize": "10px",
                  "textAnchor": "end",
                  "transform": "translate(-3px, 3px)",
                }
              }
            >
              70
            </text>
          </g>
          <g
            transform="translate(24, 275)"
          >
            <line
              stroke="#fff"
              x2="6"
            />
            <text
              style={
                {
                  "fill": "#fff",
                  "fontSize": "10px",
                  "textAnchor": "end",
                  "transform": "translate(-3px, 3px)",
                }
              }
            >
              60
            </text>
          </g>
          <g
            transform="translate(24, 342.5)"
          >
            <line
              stroke="#fff"
              x2="6"
            />
            <text
              style={
                {
                  "fill": "#fff",
                  "fontSize": "10px",
                  "textAnchor": "end",
                  "transform": "translate(-3px, 3px)",
                }
              }
            >
              50
            </text>
          </g>
          <g
            transform="translate(24, 410)"
          >
            <line
              stroke="#fff"
              x2="6"
            />
            <text
              style={
                {
                  "fill": "#fff",
                  "fontSize": "10px",
                  "textAnchor": "end",
                  "transform": "translate(-3px, 3px)",
                }
              }
            >
              40
            </text>
          </g>
          <g
            transform="translate(24, 477.49999999999994)"
          >
            <line
              stroke="#fff"
              x2="6"
            />
            <text
              style={
                {
                  "fill": "#fff",
                  "fontSize": "10px",
                  "textAnchor": "end",
                  "transform": "translate(-3px, 3px)",
                }
              }
            >
              30
            </text>
          </g>
          <g
            transform="translate(24, 545)"
          >
            <line
              stroke="#fff"
              x2="6"
            />
            <text
              style={
                {
                  "fill": "#fff",
                  "fontSize": "10px",
                  "textAnchor": "end",
                  "transform": "translate(-3px, 3px)",
                }
              }
            >
              20
            </text>
          </g>
          <g
            transform="translate(24, 612.5)"
          >
            <line
              stroke="#fff"
              x2="6"
            />
            <text
              style={
                {
                  "fill": "#fff",
                  "fontSize": "10px",
                  "textAnchor": "end",
                  "transform": "translate(-3px, 3px)",
                }
              }
            >
              10
            </text>
          </g>
          <g
            transform="translate(24, 680)"
          >
            <line
              stroke="#fff"
              x2="6"
            />
            <text
              style={
                {
                  "fill": "#fff",
                  "fontSize": "10px",
                  "textAnchor": "end",
                  "transform": "translate(-3px, 3px)",
                }
              }
            >
              0
            </text>
          </g>
        </g>
        <path />
        <g>
          <path
            className=""
          />
        </g>
        <g>
          <path
            className=""
          />
        </g>
        <g>
          <path
            className=""
          />
        </g>
        <g>
          <path
            className=""
          />
        </g>
        <g>
          <path
            className=""
          />
        </g>
        <g>
          <path
            className="draw-line"
          />
        </g>
        <g>
          <circle
            className="plot"
            cx={10.199999999999998}
            cy={599}
            fill="#38bdf8"
            r={6}
            stroke="#fff"
          />
          <circle
            className="plot"
            cx={0}
            cy={545}
            fill="#38bdf8"
            r={6}
            stroke="#fff"
          />
          <circle
            className="plot"
            cx={-11.999999999999998}
            cy={511.25}
            fill="#38bdf8"
            r={6}
            stroke="#fff"
          />
          <circle
            className="plot"
            cx={-27}
            cy={484.24999999999994}
            fill="#38bdf8"
            r={6}
            stroke="#fff"
          />
          <circle
            className="plot"
            cx={-42}
            cy={362.75000000000006}
            fill="#38bdf8"
            r={6}
            stroke="#fff"
          />
          <circle
            className="plot"
            cx={-72}
            cy={207.50000000000003}
            fill="#38bdf8"
            r={6}
            stroke="#fff"
          />
          <circle
            className="plot"
            cx={-90}
            cy={153.5}
            fill="#38bdf8"
            r={6}
            stroke="#fff"
          />
          <circle
            className="plot"
            cx={-108}
            cy={72.49999999999999}
            fill="#38bdf8"
            r={6}
            stroke="#fff"
          />
        </g>
      </svg>
      <h4
        className="text-xs text-sky-500 text-center mt-1"
      >
        Age (in months)
      </h4>
    </div>
  </div>
</div>
`;
